File: apps/ios/Barns/BarnsApp/Presentation/Support/SupportViewModel.swift
import Combine

@MainActor
final class SupportViewModel: ObservableObject {
    enum State: Equatable {
        case loading
        case loaded(CompanyInfo)
        case failed(String)
    }

    @Published private(set) var state: State = .loading

    private let getSupportInfoUseCase: GetSupportInfoUseCase

    init(getSupportInfoUseCase: GetSupportInfoUseCase) {
        self.getSupportInfoUseCase = getSupportInfoUseCase
    }

    func load() async {
        state = .loading
        do {
            let info = try await getSupportInfoUseCase.execute()
            state = .loaded(info)
        } catch {
            state = .failed("Unable to load support info. Please try again.")
        }
    }
}

File: apps/android/app/src/main/java/com/barns/app/presentation/support/SupportViewModel.kt
package com.barns.app.presentation.support

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.barns.app.domain.model.CompanyInfo
import com.barns.app.domain.usecase.support.GetSupportInfoUseCase
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class SupportViewModel(
    private val getSupportInfoUseCase: GetSupportInfoUseCase,
) : ViewModel() {
    sealed interface State {
        object Loading : State
        data class Loaded(val info: CompanyInfo) : State
        data class Failed(val message: String) : State
    }

    private val _state = MutableStateFlow<State>(State.Loading)
    val state: StateFlow<State> = _state.asStateFlow()

    fun load() {
        _state.value = State.Loading
        viewModelScope.launch {
            runCatching { getSupportInfoUseCase.execute() }
                .onSuccess { info -> _state.value = State.Loaded(info) }
                .onFailure { _state.value = State.Failed("Unable to load support info. Please try again.") }
        }
    }
}
